rust/rust.cc

```diff
@@ -0,0 +1,231 @@
+    uint64_t npnr_pipid_null() { return wrap(PipId()); }
+
+    int npnr_context_get_grid_dim_x(const Context *ctx) { return ctx->getGridDimX(); }
+    int npnr_context_get_grid_dim_y(const Context *ctx) { return ctx->getGridDimY(); }
+    void npnr_context_bind_bel(Context *ctx, uint64_t bel, CellInfo* cell, PlaceStrength strength) { return ctx->bindBel(unwrap_bel(bel), cell, strength); }
+    void npnr_context_unbind_bel(Context *ctx, uint64_t bel) { return ctx->unbindBel(unwrap_bel(bel)); }
+    bool npnr_context_check_bel_avail(Context *ctx, uint64_t bel) { return ctx->checkBelAvail(unwrap_bel(bel)); }
+    void npnr_context_bind_wire(Context *ctx, uint64_t wire, NetInfo* net, PlaceStrength strength) { ctx->bindWire(unwrap_wire(wire), net, strength); }
+    void npnr_context_unbind_wire(Context *ctx, uint64_t wire) { ctx->unbindWire(unwrap_wire(wire)); }
+    void npnr_context_bind_pip(Context *ctx, uint64_t pip, NetInfo* net, PlaceStrength strength) { ctx->bindPip(unwrap_pip(pip), net, strength); }
+    void npnr_context_unbind_pip(Context *ctx, uint64_t pip) { ctx->unbindPip(unwrap_pip(pip)); }
+    uint64_t npnr_context_get_pip_src_wire(const Context *ctx, uint64_t pip) { return wrap(ctx->getPipSrcWire(unwrap_pip(pip))); }
+    uint64_t npnr_context_get_pip_dst_wire(const Context *ctx, uint64_t pip) { return wrap(ctx->getPipDstWire(unwrap_pip(pip))); }
+    float npnr_context_estimate_delay(const Context *ctx, uint64_t src, uint64_t dst) { return ctx->getDelayNS(ctx->estimateDelay(unwrap_wire(src), unwrap_wire(dst))); }
+    float npnr_context_get_pip_delay(const Context *ctx, uint64_t pip) { return ctx->getDelayNS(ctx->getPipDelay(unwrap_pip(pip)).maxDelay()); }
+    float npnr_context_get_wire_delay(const Context *ctx, uint64_t wire) { return ctx->getDelayNS(ctx->getWireDelay(unwrap_wire(wire)).maxDelay()); }
+    float npnr_context_delay_epsilon(const Context *ctx) { return ctx->getDelayNS(ctx->getDelayEpsilon()); }
+    Loc npnr_context_get_pip_location(const Context *ctx, uint64_t pip) { return ctx->getPipLocation(unwrap_pip(pip)); }
+    bool npnr_context_check_pip_avail_for_net(const Context *ctx, uint64_t pip, NetInfo *net) { return ctx->checkPipAvailForNet(unwrap_pip(pip), net); }
+
+    uint64_t npnr_context_get_pips_leak(const Context *ctx, uint64_t **pips) {
+        auto size = size_t{};
+        for (auto _pip : ctx->getPips())
+            size++;
+        auto *pip_vec = new std::vector<uint64_t>{};
+        pip_vec->reserve(size);
+        for (auto pip : ctx->getPips())
+            pip_vec->push_back(wrap(pip));
+        *pips = pip_vec->data();
+        // Yes, by never deleting pip_vec, we leak memory.
+        return size;
+    }
+
+    uint64_t npnr_context_get_wires_leak(const Context *ctx, uint64_t **wires) {
+        auto size = size_t{};
+        for (auto _wire : ctx->getWires())
+            size++;
+        auto *wire_vec = new std::vector<uint64_t>{};
```

Comment:
I think this can be simplified yet further to

`uint64_t *wires = new uint64_t[size];`